Sources/Geno/Geno.swift

```
@@ -0,0 +1,377 @@
+            "\(parameter.name).asField"
+        } else {
+            "\(parameterToFieldType(parameter, allowSchema: true))(\(parameter.name))"
+        }
+    }.joined(separator: ", ")
+}
+
+func functionParameters(f: Contract.ABI.Function) -> [String] {
+    return f.inputs.map { "\($0.name): \(try! typeMapper(for: $0.internalType))" }
+}
+
+func returnValue(f: Contract.ABI.Function) -> String {
+    f.outputs.map { try! typeMapper(for: $0.internalType) }.joined(separator: ", ")
+}
+
+func parameterToFieldType(_ parameter: Contract.ABI.Function.Parameter, allowSchema: Bool = false) -> String {
+    let inner: String
+    let baseParameter = baseParameter(parameter)
+
+    if isTuple(parameter) {
+        if allowSchema, let structName = structName(parameter) {
+            inner = "\(structName).schema"
+        } else {
+            let componentTypes = parameter.components?.map { parameterToFieldType($0, allowSchema: true) } ?? []
+            inner = ".tuple([\(componentTypes.joined(separator: ", "))])"
+        }
+
+    } else {
+        inner = ".\(baseParameter.type)"
+    }
+
+    if isArray(parameter) {
+        return ".array(\(inner))"
+    } else {
+        return inner
+    }
+}
+
+func parameterToMatchableFieldType(p: Contract.ABI.Function.Parameter, index: Int, asField: Bool = false, isChild: Bool = false) -> String {
+    // Note: it's comical that all of these cases need to be handled differently...
+    let varName = parameterVar(parameter: p, index: index, withPrefix: "var")
+
+    if isArray(p) {
+        let baseParameter = baseParameter(p)
+        if isTuple(p) {
+            if let structName = structName(p) {
+                if asField {
+                    return ".array(\(structName).schema, \(varName).map { $0.asField })"
+                } else {
+                    return ".array(\(structName).schema, \(varName))"
+                }
+            } else {
+                return "tuplearray"
```

Can we raise a fatal error here, just so as not to actually generate in this case that we think is impossible.